daemon/events/events.go

```diff
@@ -0,0 +1,144 @@
+func NewEvents() *Events {
+	return &Events{
+		broadcaster: goevents.NewBroadcaster(),
+	}
+}
+
+// Publish sends an event. The caller will be considered the initial
+// publisher of the event. This means the timestamp will be calculated
+// at this point and this method may read from the calling context.
+// TODO: do we need lock when use broadcaster???
+func (e *Events) Publish(ctx context.Context, action string, eventType types.EventType, actor *types.EventsActor) error {
+	var err error
+	// ensure actor not nil
+	if actor == nil {
+		actor = &types.EventsActor{}
+	}
+
+	now := time.Now().UTC()
+	msg := types.EventsMessage{
+		Action:   action,
+		Type:     eventType,
+		Actor:    actor,
+		Time:     now.Unix(),
+		TimeNano: now.UnixNano(),
+	}
+
+	// compatibility with moby
+	switch eventType {
+	case types.EventTypeContainer:
+		msg.ID = actor.ID
+		msg.Status = action
+		if actor.Attributes != nil {
+			if image, ok := actor.Attributes["image"]; ok {
+				msg.From = image
+			}
+		}
+	case types.EventTypeImage:
+		msg.ID = actor.ID
+		msg.Status = action
+	}
+
+	defer func() {
+		if err != nil {
+			logrus.Errorf("failed to publish event {action: %s, type: %s, id: %s}: %v", msg.Action, msg.Type, msg.ID, err)
+		} else {
+			logrus.Debugf("event published event {action: %s, type: %s, id: %s}", msg.Action, msg.Type, msg.ID)
+		}
+	}()
+
+	err = e.broadcaster.Write(&msg)
+	return err
+}
+
+// Subscribe to events on the Events. Events are sent through the returned
+// channel ch. If an error is encountered, it will be sent on channel errs and
+// errs will be closed. To end the subscription, cancel the provided context.
+//
+// Zero or more filters may be provided as Args. Only events that match
+// *any* of the provided filters will be sent on the channel. The filters use
+// the standard containerd filters package syntax.
+func (e *Events) Subscribe(ctx context.Context, ef *Filter) (ch <-chan *types.EventsMessage, errs <-chan error) {
+	var (
+		evch                  = make(chan *types.EventsMessage)
+		errq                  = make(chan error, 1)
+		channel               = goevents.NewChannel(0)
+		queue                 = goevents.NewQueue(channel)
+		dst     goevents.Sink = queue
+	)
+
+	closeAll := func() {
+		defer close(errq)
+		defer e.broadcaster.Remove(dst)
+		defer queue.Close()
+		defer channel.Close()
+	}
+
+	ch = evch
+	errs = errq
+
+	// add filters for event messages
+	if ef != nil && ef.filter.Len() > 0 {
+		dst = goevents.NewFilter(queue, goevents.MatcherFunc(func(gev goevents.Event) bool {
+			// TODO(ziren): maybe we need adaptor here
+			msg := gev.(*types.EventsMessage)
+			return ef.Match(*msg)
+		}))
+	}
+
+	e.broadcaster.Add(dst)
+
+	go func() {
+		defer closeAll()
+
+		var err error
+	loop:
+		for {
+			select {
+			case ev := <-channel.C:
+				env, ok := ev.(*types.EventsMessage)
+				if !ok {
+					err = errors.Errorf("invalid message encountered %#v; please file a bug", ev)
```

Comment:
`please file a bug`. funny. If we meet it, could we panic?

Reply:
no need panic, return err is ok, WDYT?

Reply:
ok